opt_einsum/contract.py

```diff
@@ -438,9 +439,15 @@ def _core_contract(operands, contraction_list, backend='numpy', **einsum_kwargs)
     # Start contraction loop
     for num, contraction in enumerate(contraction_list):
         inds, idx_rm, einsum_str, remaining, blas_flag = contraction
-        tmp_operands = []
-        for x in inds:
-            tmp_operands.append(operands.pop(x))
+
+        # check if we are performing the pre-pass of an expression with constants
+        if evaluate_constants:
+            # if so, keep contracting until we find any non-constant operators, then break out
+            operands_copy = list(operands)
+            if any(operands_copy.pop(x) is None for x in inds):
```

Comment:
I think my only comment is still here and if it is possible to convert this to set logic.

Reply:
I'm not sure about set logic (if you have a snippet in mind please share!) but I did simplify this bit, based on realising that ``inds`` are actually always reverse-sorted so no popping needs to be done.